python/xorbits/_mars/storage/mmap.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2022-2023 XProbe Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import asyncio
+import mmap
+import warnings
+from typing import Dict, List, Tuple, Union
+
+import cloudpickle
+from xoscar.serialization import deserialize
+from xoscar.serialization.aio import BUFFER_SIZES_NAME, get_header_length
+
+from ..utils import implements
+from .base import ObjectInfo, StorageBackend, StorageLevel, register_storage_backend
+from .core import StorageFileObject
+from .filesystem import DiskStorage
+
+
+@register_storage_backend
+class MMAPStorage(DiskStorage):
+    name = "mmap"
+    is_seekable = True
+
+    def __init__(self, **kw):
+        DiskStorage.__init__(self, **kw)
+
+    @classmethod
+    @implements(StorageBackend.setup)
+    async def setup(cls, **kwargs) -> Tuple[Dict, Dict]:
+        disk_init, disk_teardown = await DiskStorage.setup(**kwargs)
+
+        return disk_init, disk_teardown
+
+    @staticmethod
+    @implements(StorageBackend.teardown)
+    async def teardown(**kwargs):
+        fs = kwargs.get("fs")
+        root_dirs = kwargs.get("root_dirs")
+        try:
+            for d in root_dirs:
+                fs.delete(d, recursive=True)
+        except:  # pragma: no cover
+            warnings.warn(
+                f"Unexpected exceptions occur when deleting directories, "
+                f"please delete these directories manually: {root_dirs}"
+            )
+
+    @property
+    @implements(StorageBackend.level)
+    def level(self) -> StorageLevel:
+        return StorageLevel.MEMORY
+
+    @property
+    @implements(StorageBackend.size)
+    def size(self) -> Union[int, None]:
+        return self._size
+
+    @staticmethod
+    def _get_file_via_mmap(object_id: str):
+        with open(object_id, "r+b") as f:
+            mm = mmap.mmap(f.fileno(), 0)
```

Comment:
Use a context manager instead `with mmap.mmap(...) as f`.

Reply:
Now use memoryview to get slice, then cannot use context manager way.